### ISSUE TITLE
Contribution & git workflow

### ISSUE THREAD

**Issue by Somotaw** [TooAngel/screeps#104]
# My edited suggestion :)

# Contribution
All kind of contribution is welcome, issues, contact via channels, pull requests.

## Priority
 - Make contribution as easy as possible
 - Easy to understand contribution ideas
 - History looking good, would be nice

## Flow

- Start by forking `TooAngel/screeps`
- Make your own branch off the latest master
- Start a pull request with the prefix [WIP] if you want ongoing reviews during development
- Work on your feature until you think its ready push your commits if you want ongoing reviews
- Rebase your whole feature interactively, using squash or fixup on all merge commits you have made and anything else you would like to edit
- Rebase your feature off latest master
- Test your feature or have someone test it for you, push your commits if you want reviews
- If master has been updated start again from the first `Rebase` step
- Run `npm install` in your repository if you havent already done so, then `grunt dev`
- Push commits
- Remove [WIP] prefix from your pull request
- Await reviews
- If there are issues, prefix the pull request again and start again from first rebase step
- Await merge to master and pray everything works \o/
- Remove the feature branch

# Details

### There is only one eternal branch - master
- All other branches (feature, hotfix, and whatever else you need) are temporary and only used as a convenience to share code with other developers and as a backup measure. They are always removed once the changes present on them land on master.

### Features are integrated onto the master branch without having any merge commits in them
- This ensures linear history.

### Pull Request

- Small PRs are more welcome than big ones and will be merged faster, especially because of the review time. 
- One PR should preferrably only tackle one feature.

### Commits
- One commit should fully tackle one topic, not more not less.

### Work in progress

- You are fine to do on your branch what ever you want. Especially rebasing or squashing before removing the WIP label is expected.
- `--fixup` commits are preferred if you need to commit something that handles the same topic as another one of your commits in your feature or hotfix. These are wither done with the --fixup option for git cli or you can start your commit message with fixup! so they can be combined automatically with the previous or targeted commit on merge or rebase.

## What to do
Have a peak at: [Issues](https://github.com/TooAngel/screeps/issues?q=is%3Aissue+is%3Aopen+label%3Aenhancement)
These topics are open for discussion, jump in, and also for implementation. The description will reflect the latest status of the
discussion and should end up in the documentation, when finishing the
implementation.

**Comment by Somotaw** [comment]
links here for more in detail.
http://endoflineblog.com/gitflow-considered-harmful
http://endoflineblog.com/follow-up-to-gitflow-considered-harmful

**Comment by Somotaw** [comment]
instead of 

 in order to reach the state of the last release in the code, you have to look at tags and choose the latest one, while in GitFlow you just did git checkout master.

Solution if you want to have similar functionality:

You create an additional branch, you can call it 'current'
whenever you do a hotfix or a release, you need to execute: git merge <new_tag_name> while on the 'current' branch, which simply fast-forwards this branch to the newly created production version. This way, 'current' can always be checked out to reach the latest production release.

**Comment by TooAngel** [comment]
@Somotaw Can you copy the description to a comment, so that we can discuss and update the description to the current status. Later on we can copy it to a `Contribution.md` and add it to the repository.

Will comment properly later.

**Comment by Somotaw** [comment]
Like this? or did you want a pr with the md now? so we can comment on specific parts of the description?

**Comment by TooAngel** [comment]
So my perspective on this:

I think gitflow and antigitflow mainly concentrate on git, but leave out the github PR system.

# Priority

 1. Make contribution as easy as possible
 2. Easy to understand contribution ideas
 3. History looking good, would be nice

I think the only valid way for contributing is sending PRs. I want to have this public and no hierarchy. So I don't see  it, giving someone access to mess around with branches on the main repository. I'm sending PRs as you all do, too.

So my main questions are:

 - Do we want to have intermediate branches? staging, q/a, release, ...? IMHO no, who will take it from one to the other branch? Does anyone really wants to do q/a? I don't really, I'm playing with the bot and hope that we see all errors in the review, if not we will fix them afterwards.
 - Do we want to use the github autosquash feature? On PR merge you can squash all commits to one merge commit. I'm not such a fan of it, because I'm trying to make meaningful commits in my PR, why squash them. So instead of autosquash, have a rule: 'Make proper commits, good commit message, working by itself, one topic' - and we all check on PR review that this is fine. I think this should be taken care of anyway when using an RCS.
 - How to prepare a PR? I think PR should be rather small, especially to be able to review them in a good amount of time. If not possible, it is not possible :-) If you have a lot of commits due to review, I would suggest to rebase them before merging. Rebasing is bad if multiple people work on the same branch, so the `[WIP]` rule. (Btw. I think we should switch from title prefix, to label). Also if you are using proper `git commit --fixup COMMIT_ID` I think rebasing is not so bad anymore, because everybody else can do it the same way as you did. Also rebasing on the latest master would be fine and makes the history nicer. It will happen that there are sideways in the git history but shouldn't be so bad.
 - How to merge a PR? Currently I'm the only one who is allowed to merge. I'm thinking about introducing some kind of democratic PR merge. Every contributor is weighted somehow (number of commits?). A PR is automatically merged with _threshold_ is reached, a PR is merged after time is _lower threshold_ is reached, otherwise it stays open.

**Comment by Somotaw** [comment]
# My suggestion

### Contribution & git workflow

### There is only one eternal branch - master
All other branches (feature, release, hotfix, and whatever else you need) are temporary and only used as a convenience to share code with other developers and as a backup measure. They are always removed once the changes present on them land on master.
### Features are integrated onto the master branch without ever pushing merge commits to master
This way ensure linear history.
### Releases are done similarly to in GitFlow.
You create a new branch for the release, branching off at the point in master that you decide has all the necessary features. From then on new work, aimed for the next release, is pushed to master as always, and any necessary changes are pushed to the release branch (This should not be needed). Finally, once the release is ready, you tag the top of the release branch. Then, because there is one eternal branch, there is only one way to get your release to be versioned permanently - and that is to merge the release branch into master and push that changed master. After that, all the changes that were made during the release are now part of master, and the release branch is deleted.
### Hotfixes are very similar to releases
Exception is you don't branch from an arbitrary commit on master, but from the release tag that you want to make the fix in. Again, work on master continues as always, and the necessary fixes are pushed to the hotfix branch. Once the fix is ready, the procedure is exactly the same as for a release - tag the top of the branch creating a new release, merge it into master, then delete the hotfix branch.

#### With gitflow
http://endoflineblog.com/assets/gitflow-mess-cfd9aa7a4137e3e575510fcbbf38a5b6.png

#### With this model
http://endoflineblog.com/assets/antigitflow-order-66c1dff2afdcee7c315a664ab6d7c9b9.png


#### Issues
In order to reach the state of the last release in the code, you have to look at tags and choose the latest one, while in GitFlow you just did git checkout master.

#### Solution if you want to have similar functionality

You create an additional branch, you can call it 'current'
whenever you do a hotfix or a release, you need to execute: git merge <new_tag_name> while on the 'current' branch, which simply fast-forwards this branch to the newly created production version. This way, 'current' can always be checked out to reach the latest production release.


### More in detail
http://endoflineblog.com/gitflow-considered-harmful
http://endoflineblog.com/follow-up-to-gitflow-considered-harmful

**Comment by Somotaw** [comment]
without ever pushing merge commits to master mean that all merge commits to features will be ammended but not the merge to master.

**Comment by TooAngel** [comment]
IMHO most of the points antigitflow proposes doesn't fit for use:
 - We do not have any branches, except of the eternal master
 - Merge is done via PR merge commit (I really don't think we should do something else than PRs)
 - Hotfixes in our case are exactly like releases, they can only be fixed via PR

**Comment by Somotaw** [comment]
branches would constitute the branches on the forks, but work the same otherwise. Im not proposing doing anything other than pr's, the pr's will be linked to the feature branches on the forks :)

**Comment by Somotaw** [comment]
no merge commits pushed to master means rebase all merge commits to your feature before pr merg into master

**Comment by Somotaw** [comment]
I see all branches of the forks when I have the remotes added just as if it was 1 repository with branches.

**Comment by TooAngel** [comment]
Proposal for our rules:
# Contribution
All kind of contribution is welcome, issues, contact via channels, pull requests.

## Priority
 - Make contribution as easy as possible
 - Easy to understand contribution ideas
 - History looking good, would be nice

## Pull Request
Contribution is done via Pull Request.
If you are working on a feature and want to show your progress or idea, label if with WIP. This PR will not be merged, but maybe reviewed already. Also you are fine to do on your branch what ever you want. Especially rebasing on squashing before removing the WIP label is expected.
Small PRs are more welcome than big ones and will be merged faster, especially because of the review time.

## Commits
One commit should fully tackle one topic, not more not less.

## What to do
Have a peak at: [Issues](https://github.com/TooAngel/screeps/issues?q=is%3Aissue+is%3Aopen+label%3Aenhancement)
These topics are open for discussion, jump in, and also for implementation. The description will reflect the latest status of the
discussion and should end up in the documentation, when finishing the
implementation.

**Comment by mschultheiss83** [comment]
maybe some words to codestyle?
@TooAngel i miss this whitespaces a lot at: https://github.com/TooAngel/screeps/pull/77/commits/b5d97d4b26a6c1faf11eb2b93e0ecfd3c9138d16

**Comment by Somotaw** [comment]
We just need to include running grunt dev before removing WIP :)

**Comment by Somotaw** [comment]
Will fix your style :)

**Comment by Somotaw** [comment]
Working on updated version as we speak

**Comment by TooAngel** [comment]
> @TooAngel i miss this whitespaces a lot at: b5d97d4

I'm not religious about code style as long as something formats and validates the code for me. I have a slight preference against the whitespace, but if you find enough contributors to vote for this change in style (maybe send PR with the change in the config file and discuss there), I'm neutral on this :-)

**Comment by Somotaw** [comment]
I have my ide auto format on save :)

**Comment by Somotaw** [comment]
# My edited suggestion :)

# Contribution
All kind of contribution is welcome, issues, contact via channels, pull requests.

## Priority
 - Make contribution as easy as possible
 - Easy to understand contribution ideas
 - History looking good, would be nice

## Flow

- Start by forking TooAngel/screeps
- Make your own feature/yourfeaturename branch off the latest master
- Start a pull request with the prefix [WIP] if you want ongoing reviews during development
- Work on your feature until you think its ready push your commits if you want ongoing reviews
- Rebase your whole feature interactively, using squash or fixup on all merge commits you have made and anything else you would like to edit
- Rebase your feature off latest master
- Test your feature or have someone test it for you, push your commits if you want reviews
- If master has been updated start again from the first rebase step
- Push commits
- Remove [WIP] prefix from your pull request
- Await reviews
- If there are issues, prefix the pull request again and start again from first rebase step
- Await merge to master and pray everything works \o/

# Details

### There is only one eternal branch - master
- All other branches (feature, release, hotfix, and whatever else you need) are temporary and only used as a convenience to share code with other developers and as a backup measure. They are always removed once the changes present on them land on master.

### Features are integrated onto the master branch without having any merge commits in them
- This ensures linear history.

### Pull Request

- Small PRs are more welcome than big ones and will be merged faster, especially because of the review time. 
- One PR should preferrably only tackle one feature.

### Commits
- One commit should fully tackle one topic, not more not less.

### Hotfix
- You create a new branch hotfix/featurename for the hotfix, branching off at the latest point in master.
- Work is continued from "Start a pull request" in the flow list

### Work in progress

- You are fine to do on your branch what ever you want. Especially rebasing or squashing before removing the WIP label is expected.
- `--fixup` commits are preferred if you need to commit something that handles the same topic as another one of your commits in your feature or hotfix. These are wither done with the --fixup option for git cli or you can start your commit message with fixup! so they can be combined automatically with the previous or targeted commit on merge or rebase.

## What to do
Have a peak at: [Issues](https://github.com/TooAngel/screeps/issues?q=is%3Aissue+is%3Aopen+label%3Aenhancement)
These topics are open for discussion, jump in, and also for implementation. The description will reflect the latest status of the
discussion and should end up in the documentation, when finishing the
implementation.

**Comment by TooAngel** [comment]
> Rebase your whole feature interactively, using squash or fixup on all merge commits you have made and anything else you would like to edit
> Rebase your feature off latest master

I think you shouldn't rebase while developing. I think this should only be done when the WIP is removed.

**Comment by Somotaw** [comment]
well it is basically when WIP is removed unless there are other pul lrequests merged to master in that small timeframe :)

**Comment by TooAngel** [comment]
Not sure if we have to address hotfix, but I'm fine

**Comment by Somotaw** [comment]
Will make all messages very clear on slack too :) with the hitfix/feature names

**Comment by Somotaw** [comment]
Hmm im going to edit it so you should prefix your fork with username, otherwise we dont know which persons feature or hotfix is being updated if they are working on the same one when we see messages on slack

**Comment by Somotaw** [comment]
Unless you have any other thoughts? :)

**Comment by Somotaw** [comment]
Maybe add somethign about installing npm and grunt-cli? :)

**Comment by TooAngel** [comment]
Hm, not sure about making too much rules about naming branches. If I would work on another project I'm not sure if I would take care of.

`npm` and `grunt` are mentioned in the README, would keep it there.

**Comment by Somotaw** [comment]
Updated, better now? 👍

**Comment by Somotaw** [comment]
Autosquash I agree should not be used :)
Only issue not addressed now is the automatic pr merge, is that even possible without setting up our own script for it? :)

**Comment by Somotaw** [comment]
It's a good idea, especially if you want to have some time off or similar.

**Comment by TooAngel** [comment]
I have an idea for the github handling, wanted to check some things first, but will tell you later.

**Comment by Somotaw** [comment]
Okay, hope it works out :)

**Comment by Somotaw** [comment]
I have found a solution to our problems :D Simple steps will ensue below:

# Create a new organization, import repo and host it there.
# Change settings so they look like this:
![image](https://cloud.githubusercontent.com/assets/6491698/21794792/da16c3f2-d6fd-11e6-884d-9b7566ec01de.png)
![image](https://cloud.githubusercontent.com/assets/6491698/21794904/b9b2f68e-d6fe-11e6-80c6-83556f5916e5.png)
![image](https://cloud.githubusercontent.com/assets/6491698/21794919/d66d47c0-d6fe-11e6-9e24-35ac89ce0495.png)
![image](https://cloud.githubusercontent.com/assets/6491698/21794977/41ba83b2-d6ff-11e6-81cd-904da7444cf3.png)
![image](https://cloud.githubusercontent.com/assets/6491698/21794955/1b2d5b2a-d6ff-11e6-99e3-accfbb847eec.png)
![image](https://cloud.githubusercontent.com/assets/6491698/21794958/1fdce65e-d6ff-11e6-9ed7-77079baad144.png)

# Make a Team called Review Team
![image](https://cloud.githubusercontent.com/assets/6491698/21795082/fdefea54-d6ff-11e6-8dab-5cc16aa47ff3.png)
# Add repository for review team
![image](https://cloud.githubusercontent.com/assets/6491698/21795102/11f899c4-d700-11e6-947b-7be13e0bc478.png)
# ensure write access
![image](https://cloud.githubusercontent.com/assets/6491698/21795112/249739aa-d700-11e6-9f14-5f9353a34658.png)
# Add fitting users to team and organization
![image](https://cloud.githubusercontent.com/assets/6491698/21795125/4336f04e-d700-11e6-9083-b7867e5f6d80.png)
# Add permission for Review Ninja on acc
![image](https://cloud.githubusercontent.com/assets/6491698/21795328/77fec0e4-d701-11e6-9535-b70533d03c3c.png)
# Organizsation access at the bottom
![image](https://cloud.githubusercontent.com/assets/6491698/21795336/87ff25c4-d701-11e6-9f9a-cc0135101664.png)
# add the repository with the + sign next to the repository headline
![image](https://cloud.githubusercontent.com/assets/6491698/21795363/baa8ba94-d701-11e6-9580-5f3d88e181e7.png)
# Add the amount of official upvotes required in the repository settings on ReviewNinja
## upper right corner
![image](https://cloud.githubusercontent.com/assets/6491698/21795398/f4ca5804-d701-11e6-87e2-a6e1d651c18a.png)
## Set the number and the team associated with it here
![image](https://cloud.githubusercontent.com/assets/6491698/21795421/16aa5e06-d702-11e6-9307-6c9d02ee3d3a.png)
# This is where you add ninja start and can also use to request changes/fix them and comments etc
![image](https://cloud.githubusercontent.com/assets/6491698/21795468/64e2e7e6-d702-11e6-9199-7fd746915187.png)
## There are some shortcuts where you can comment with thumbs up or similar as an official review team member and it will automatically add the ninja star, thoug this seems to not always be working
## Sometimes it can look like you can press the merge button even if you cant merge, this does not work - tried and tested

# If you'd like
Convert someone to team maintainer, they can:
- Change the team's name and description
- Change the team's visibility
- Add organization members to the team
- Remove organization members from the team
- Promote an existing team member to team maintainer
- Remove the team's access to repositories
- Reinstate a former organization member
note the difference between team and member
![image](https://cloud.githubusercontent.com/assets/6491698/21795163/899bfb1a-d700-11e6-8272-dca0ed7f1ee7.png)

# Non Organizations members work just as they do right now with regarding permissions



# Extra Changes
## With this system you could also enable write acces to everyone that is in the organization giving them the following permissions
- merge pull requests that have the required amount of official review approvals
- edit branches other than master on repo, even pushing ( we could keep cooperative projects in main repo branches and personal ones where theres only one coder working on a feature in personal forks, so they cant be disturbed unless they check the allow maintainers to edit box, which will allow anyone with write(not not push) access to master to also edit the branch in the personal fork
## We could also enable those with write access to create new repositories in the organization
- But im not sure what that would be good for unless we somehow use a repository for a big feature instead of just a branch

# Added Bonus!
## Anyone with write acces can edit things like labels and assignments and projects making them valid options for us, zenhub too as you cant move issues between pipes without write access. This would still work for official Review Team only if not enabled for all Organization Members

# Before you ask, this will not work with a personal repository as you can not make a team to assign as official reviewers and any approved reviews in review ninja will count towards the number required giving anyone ability to approve a PR, If they also have write acces this means that they can approve + pull.

**Comment by Somotaw** [comment]
phew that took alot of time to writeup haha, hope its as clear as intended 💃

**Comment by Somotaw** [comment]
you will still be owner and have same permissions you have now :)
